app/models/group_assignment.rb

```diff
@@ -0,0 +1,21 @@
+class GroupAssignment < ApplicationRecord
+  belongs_to :group_offer
+  belongs_to :volunteer
+  has_many :group_assignment_logs
+
+  after_save :save_group_assignment_logs, if: :dates_updated?
+  before_destroy :save_group_assignment_logs
+
+  def save_group_assignment_logs
+    group_assignment_logs.create!(group_offer_id: group_offer_id, volunteer_id: volunteer_id,
+      group_assignment_id: id, title: group_offer.title,
+      period_start: period_start_before_last_save, period_end: period_end_before_last_save,
+      responsible: responsible)
+  end
+
+  private
+
+  def dates_updated?
+    saved_change_to_period_start? || saved_change_to_period_end?
```

Comment:
where do these come from? or how are they defined?

Reply:
http://blog.toshima.ru/2017/04/06/saved-change-to-attribute.html